nvflare/app_common/ccwf/swarm_client_ctl.py

```diff
@@ -0,0 +1,620 @@
+
+        rc = result.get_return_code(ReturnCode.OK)
+        if rc != ReturnCode.OK:
+            self.log_error(fl_ctx, f"Bad result from {client_name} for round {result_round}: {rc}.")
+            return make_reply(ReturnCode.EXECUTION_EXCEPTION)
+
+        fl_ctx.set_prop(AppConstants.CURRENT_ROUND, self.for_round, private=True, sticky=True)
+        fl_ctx.set_prop(AppConstants.TRAINING_RESULT, result, private=True, sticky=False)
+        self.fire_event(AppEventType.BEFORE_CONTRIBUTION_ACCEPT, fl_ctx)
+
+        accepted = self.aggregator.accept(result, fl_ctx)
+        accepted_msg = "ACCEPTED" if accepted else "REJECTED"
+        self.log_info(
+            fl_ctx, f"Contribution from {client_name} {accepted_msg} by the aggregator at round {result_round}."
+        )
+
+        fl_ctx.set_prop(AppConstants.AGGREGATION_ACCEPTED, accepted, private=True, sticky=False)
+        self.fire_event(AppEventType.AFTER_CONTRIBUTION_ACCEPT, fl_ctx)
+        return make_reply(ReturnCode.OK)
+
+    def aggregate(self):
+        fl_ctx = self.fl_ctx
+        self.log_info(fl_ctx, f"Start aggregation for round {self.for_round}")
+        self.fire_event(AppEventType.BEFORE_AGGREGATION, fl_ctx)
+        aggr_result = self.aggregator.aggregate(fl_ctx)
+        fl_ctx.set_prop(AppConstants.AGGREGATION_RESULT, aggr_result, private=True, sticky=False)
+        self.fire_event(AppEventType.AFTER_AGGREGATION, fl_ctx)
+        self.log_info(fl_ctx, f"Finished aggregation for round {self.for_round}")
+
+        mine_is_better = False
+        if self.current_best_global_metric is not None:
+            if self.executor.best_metric is not None and self.executor.best_metric > self.current_best_global_metric:
+                mine_is_better = True
+        elif self.executor.best_metric is not None:
+            mine_is_better = True
+
+        if mine_is_better:
+            self.log_info(
+                fl_ctx, f"I got better metric {self.executor.best_metric} at round {self.executor.best_round}"
+            )
+            best_round = self.executor.best_round
+            best_metric = self.executor.best_metric
+            best_client = self.executor.me
+        else:
+            best_round = self.current_best_round
+            best_metric = self.current_best_global_metric
+            best_client = self.current_best_client
+
+        self.log_info(fl_ctx, f"global best metric is {best_metric} from client {best_client} at round {best_round}")
+
+        aggr_result.set_header(Constant.ROUND, best_round)
+        aggr_result.set_header(Constant.METRIC, best_metric)
+        aggr_result.set_header(Constant.CLIENT, best_client)
+
+        return aggr_result
+
+    def is_done(self):
+        unfinished = 0
+        for c, s in self.trainer_statuses.items():
+            if not s.reply_time:
+                unfinished += 1
+        if unfinished == 0:
+            return True
+
+        # timeout?
+        now = time.time()
```

Comment:
Not needed since the calling logic doesn't care about the reason.
But it might be a good idea to have a log message for different abnormal DONEs. We could even add stats to pools here (later).

Reply:
Added log messages for abnormal DONEs.